src/mesh/interpolation.cxx

```diff
@@ -63,161 +67,142 @@ const Field3D interp_to(const Field3D &var, CELL_LOC loc)
+            ( (&var.yup() != &var) || (&var.ydown() != &var))) {
+          // Field "var" has distinct yup and ydown fields which
+          // will be used to calculate a derivative along
+          // the magnetic field
+          throw BoutException("Cannot interpolate in Y direction\n"
+                              " - Not enough boundary cells\n"
+                              " - at least 2 are needed.\n"
+                              " - split yup/ydown has only one guard cell!");
+        }
+        else {
+          // var has no yup/ydown fields, so we need to shift into field-aligned coordinates
+
+          Field3D var_fa = mesh->toFieldAligned(var);
+          if (mesh->ystart < 2) {
+            throw BoutException("Cannot interpolate in Y direction\n"
+                                " - Not enough boundary cells\n"
+                                " - at least 2 are needed!");
+          }
+          // More than one guard cell, so set pp and mm values
+          // This allows higher-order methods to be used
           for(const auto &i : result.region(RGN_NOY)) {
-	    // Set stencils
-	    s.c = var_fa[i];
-	    s.p = var_fa[i.yp()];
-	    s.m = var_fa[i.ym()];
-	    s.pp = var_fa[i.offset(0,2,0)];
-	    s.mm = var_fa[i.offset(0,-2,0)];
-	    
-	    if ((location == CELL_CENTRE) && (loc == CELL_YLOW)) {
-	      // Producing a stencil centred around a lower Y value
-	      s.pp = s.p;
-	      s.p  = s.c;
-	    } else if(location == CELL_YLOW) {
-	      // Stencil centred around a cell centre
-	      s.mm = s.m;
-	      s.m  = s.c;
-	    }
-	    
-	    result[i] = interp(s);
-	  }
-	  } else {
-	    // Only one guard cell, so no pp or mm values
-	    s.pp = nan("");
-	    s.mm = nan("");
-            for(const auto &i : result.region(RGN_NOY)) {
-	      // Set stencils
-	      s.c = var_fa[i];
-	      s.p = var_fa[i.yp()];
-	      s.m = var_fa[i.ym()];
-	      
-	      if ((location == CELL_CENTRE) && (loc == CELL_YLOW)) {
-		// Producing a stencil centred around a lower Y value
-		s.pp = s.p;
-		s.p  = s.c;
-	      } else if(location == CELL_YLOW) {
-		// Stencil centred around a cell centre
-		s.mm = s.m;
-		s.m  = s.c;
-	      }
-	      
-	      result[i] = interp(s);
-	    }
-	  }
-	}
-	break;
-	// Need to communicate in Y
+            // Set stencils
+            s.c = var_fa[i];
+            s.p = var_fa[i.yp()];
+            s.m = var_fa[i.ym()];
+            s.pp = var_fa[i.offset(0,2,0)];
+            s.mm = var_fa[i.offset(0,-2,0)];
+
+            if ((location == CELL_CENTRE) && (loc == CELL_YLOW)) {
+              // Producing a stencil centred around a lower Y value
+              s.pp = s.p;
+              s.p  = s.c;
+            } else if(location == CELL_YLOW) {
+              // Stencil centred around a cell centre
+              s.mm = s.m;
+              s.m  = s.c;
+            } else {
+              // We should never be here
+              throw BoutException("Maybe a butterfly defracted an X-ray?");
```

Comment:
True it would be a bug to end up here, but a more informative exception message might help fix that bug if it happens. Maybe something like 'Inconsistent values of location and loc, here one should be CELL_CENTRE and the other should be CELL_YLOW'?

Reply:
@johnomotani better?

Reply:
@dschwoerer Not really.

This if-statement can just be simplified to:
```cpp
            if (location == CELL_CENTRE) {
              // Producing a stencil centred around a lower Y value
              s.pp = s.p;
              s.p  = s.c;
            } else {
              // Stencil centred around a cell centre
              s.mm = s.m;
              s.m  = s.c;
            }
```
and similarly for the other one. The original code was overly cautious. As you noticed, there is no way, short of bit-flipping, for the code to reach that exception -- therefore it can be removed. Tidying up the other conditionals as above makes it clearer there is no third option.